tools/tpm2_checkquote.c

```diff
@@ -0,0 +1,411 @@
+            goto err;
+        }
+        if (!pcr_print_pcr_struct(&pcrSel, &pcrs)) {
+            LOG_ERR("Failed to print PCR values related to quote!");
+            goto err;
+        }
+        tpm2_tool_output("calcDigest: ");
+        tpm2_util_hexdump(ctx.pcrHash.buffer, ctx.pcrHash.size);
+        tpm2_tool_output("\n");
+    }
+
+    // Figure out the extra data (nonce) from this message
+    if (!tpm2_util_get_digest_from_quote(msg, &ctx.quoteHash, &ctx.quoteExtraData)) {
+        LOG_ERR("Failed to get digest from quote!");
+        goto err;
+    }
+
+    // Figure out the digest for this message
+    bool res = tpm2_openssl_hash_compute_data(ctx.halg, msg->attestationData, 
+        msg->size, &ctx.msgHash);
+    if (!res) {
+        LOG_ERR("Compute message hash failed!");
+        goto err;
+    }
+    tpm2_tool_output("msgDigest: ");
+    tpm2_util_hexdump(ctx.msgHash.buffer, ctx.msgHash.size);
+    tpm2_tool_output("\n");
+
+    return_value = true;
+
+err:
+    free(msg);
+    return return_value;
+
+}
+
+static bool on_option(char key, char *value) {
+
+	switch (key) {
+	case 'c':
+	    ctx.pubkey_file_path = value;
+	    break;
+	case 'G': {
+		ctx.halg = tpm2_alg_util_from_optarg(value, tpm2_alg_util_flags_hash);
+		if (ctx.halg == TPM2_ALG_ERROR) {
+			LOG_ERR("Unable to convert algorithm, got: \"%s\"", value);
+			return false;
+		}
+		ctx.flags.halg = 1;
+	}
+		break;
+	case 'm': {
+		ctx.msg_file_path = value;
+		ctx.flags.msg = 1;
+	}
+		break;
+	case 'f': {
+		ctx.format = tpm2_alg_util_from_optarg(value, tpm2_alg_util_flags_sig);
+		if (ctx.format == TPM2_ALG_ERROR) {
+		    LOG_ERR("Unknown signing scheme, got: \"%s\"", value);
+		    return false;
+		}
+
+		ctx.flags.fmt = 1;
+	} break;
+	case 'q':
+		ctx.extraData.size = sizeof(ctx.extraData) - 2;
+		if(tpm2_util_hex_to_byte_structure(value, &ctx.extraData.size, ctx.extraData.buffer) != 0)
+		{
+			LOG_ERR("Could not convert \"%s\" from a hex string to byte array!", value);
+			return false;
+		}
+		ctx.flags.extra = 1;
+		break;
+	case 's':
+		ctx.sig_file_path = value;
+		ctx.flags.sig = 1;
+		break;
+	case 'p':
+		ctx.pcr_file_path = value;
+		ctx.flags.pcr = 1;
+		break;
+		/* no default */
+	}
+
+    return true;
+}
+
+bool tpm2_tool_onstart(tpm2_options **opts) {
+
+    const struct option topts[] = {
+            { "halg",         required_argument, NULL, 'G' },
+            { "message",      required_argument, NULL, 'm' },
+            { "format",       required_argument, NULL, 'f' },
+            { "sig",          required_argument, NULL, 's' },
+            { "pcrs",         required_argument, NULL, 'p' },
+            { "pubkey",       required_argument, NULL, 'c' },
+            { "qualify-data",         required_argument, NULL, 'q' },
+    };
+
+
+    *opts = tpm2_options_new("G:m:f:s:t:c:p:q:", ARRAY_LEN(topts), topts,
+                             on_option, NULL, TPM2_OPTIONS_NO_SAPI);
```

Comment:
tools that don't need a SAPI go to tools/misc

Reply:
Tool has now been moved to tools/misc!